pkg/endpoint/endpoint.go

```diff
@@ -361,14 +397,122 @@ func (e *Endpoint) WaitForProxyCompletions() error {
 	return nil
 }
 
+// runK8sCiliumEndpointSync starts a controller that syncronises the endpoint
+// to the corresponding k8s CiliumEndpoint CRD
+// ep.Mutex must be help for the setup portion of this function, but not the
+// controller.
+// CiliumEndpoint objects follow a nodename-cep-endpointID scheme
+func (e *Endpoint) runK8sCiliumEndpointSync() {
+	var (
+		endpointID     = e.ID
+		controllerName = fmt.Sprintf("sync-to-k8s-ciliumendpoint (%v)", endpointID)
+		epName         = fmt.Sprintf("%v-cep-%v", node.GetName(), endpointID)
+		scopedLog      = e.getLogger().WithField("controller", controllerName)
+	)
+
+	if !k8s.IsEnabled() {
+		scopedLog.Debug("Not starting controller because k8s is disabled")
+		return
+	}
+
+	// The health endpoint is created outside of k8s and shouldn't create a CEP
+	// for it. It has a specific label on it and we do not start the controller
+	// in that case.
+	healthLbl := labels.NewLabel(labels.IDNameHealth, "", labels.LabelSourceReserved)
+	for _, lbl := range e.OpLabels.OrchestrationIdentity {
+		if healthLbl.Equals(lbl) {
+			scopedLog.Debug("Not starting controller because it is the reserved internal health endpoint")
+			return
+		}
+	}
+
+	ciliumClient, err := getCiliumClient()
+	if err != nil {
+		scopedLog.WithError(err).Error("Not starting controller because unable to get cilium k8s client")
+		return
+	}
+
+	var lastMdl *models.Endpoint
+
+	// NOTE: The controller functions do NOT hold the endpoint locks
+	e.controllers.UpdateController(controllerName,
+		controller.ControllerParams{
+			RunInterval: 30 * time.Second,
+			DoFunc: func() error {
+				namespace := e.GetK8sNamespace()
+				if namespace == "" {
+					scopedLog.Debug("Skipping CiliumEndpoint update because it has no k8s namespace")
+					return nil
+				}
+
+				mdl := e.GetModel()
+				if reflect.DeepEqual(mdl, lastMdl) {
+					scopedLog.Debug("Skipping CiliumEndpoint update because it has not changed")
+					return nil
+				}
+				lastMdl = mdl
+
+				cep, err := ciliumClient.CiliumEndpoints(namespace).Get(epName, meta_v1.GetOptions{})
+				switch {
+				// A real error
+				case err != nil && !k8serrors.IsNotFound(err):
+					scopedLog.WithError(err).Error("Cannot get CEP for update")
+					return err
+
+				// do an update
+				case err == nil:
+					// Update the copy of the cep
+					(*cilium_v2.CiliumEndpointDetail)(mdl).DeepCopyInto(&cep.Status)
```

Comment:
Should we check if `DeepCopyInto()` failed and left an empty cep?

Reply:
We should. The Deepcopy should never fail but there is other error indicator. +1